pitch-shift effect

xavriley opened the pull request:
```
live_loop(:s) do
  with_fx :pitch_shift, {
      window_size: 0.02,
      pitch_ratio: 0.4,
      pitch_dispersion: 0.001,
      time_dispersion: 0.01} do

    sample :loop_amen, rate: 1
    sleep sample_duration(:loop_amen, rate: 1)

  end
end
```

Adds fx for Overtone's `pitch-shift` which allows a block to have the
pitch changed independent of anything else. The docs and params need
further thought.

xavriley commented:
Creating PR for further discussion. Probably not ready to merge yet

samaaron commented:
Looks like it's on the right track to me. It's missing the actual compiled synthdef though. How does it sound and is it fun to play with?

xavriley commented:
Very fun, although the params need to be quite different depending on the input. e.g. to shift `loop_amen` without horrible comb artifacts it needs quite small values for the params (which is what I've done in this PR)

https://www.dropbox.com/s/l83287o0imghoit/amen-pitch-shift.wav?dl=0

However playing with a synth line, it sounds very bitcrushy. It does sound quite fun when you make the window size and dispersions large but it's a different sort of effect:

https://www.dropbox.com/s/uts5bd5rhthz2hl/pitch-shift-gangnam.wav?dl=0

Could easily confuse users I think.

samaaron commented:
Sounds good to me. It's an FX and like all other FX people need to play with the params to find sounds they like. I just think we need to communicate that you have to tweak around to get the desired effect and use it as a learning opportunity to explain things like window size etc.

xavriley commented:
I've changed the pitch_ratio to take a midi semitone instead. I think that's going to be more intuitive and useful to kids in school etc.

I've also added docs that try to describe how it works and how to get the best sound. I've also adjusted the params to work better for pitched sounds (like `:guit_em9`) in the default case as I think that's what people would expect. The window size would need to be dropped for something like `:loop_amen` to sound good but I have said as much in the docstrings.

todo:

  - [] Needs testing on the RPi
  - [] Need to decide on whether to allow slide/modulation on any of the params.

samaaron commented:
Also need to test on windows. I wonder if its this ugen causing the growl issues. Also definitely allow slide paeans in any params that can be modified safely.

xavriley commented:
Runs fine on my RPi model B

xavriley commented:
I can confirm sliding works a treat on OSX and RPi.

Todo

- [ ] Test on windows

samaaron commented:
@jweather Could you have a play with this on Windows?

samaaron commented:
@xavriley, have you tried using really obscure and weird param values to see if it breaks things? Some ugens are known to crash scsynth with specific params - others just totally crap the system out and sound evil!

xavriley commented:
It's fine with crazy params (1000000, 0.000000001 etc.) for me.

I've limited the lower bound of pitch to -72 as that is roughly where it stops becoming audible and probably more range than you'd need musically speaking.

I did come across an issue with live_loop swallowing the out of bounds validation which I'll file in a separate issue.

jweather commented:
Xav's :guit_em9 sample is suitably warbly on Windows.  The example at the top of this PR doesn't do any pitch shifting, did some of the args change?

samaaron commented:
@jweather, `pitch_ratio` has been converted to `pitch` which now represents MIDI notes rather than a rate multiplier.

samaaron commented:
@jweather could you also stress test it?

jweather commented:
`mix` arg doesn't seem to do anything, should it?

xavriley commented:
`mix` isn't used - thanks for catching that! I'll push another version without my lazy copy-and-paste mistakes

xavriley commented:
Ready to merge methinks

samaaron commented:
Thanks @xavriley :-)
